Build failure with libstdc++ debug iterators—iterator type mismatch

### What happened?

When building nlopt with `-D_GLIBCXX_DEBUG`, i.e. libstdc++'s debug iterators, solver.cc fails to build. This happens because there is an iterator type mismatch that happens to disappear in libstdc++ when not using debug iterators. Specifically:

https://github.com/stevengj/nlopt/blob/b22a9f7f65e589022bc9a775ad9977c792fab123/src/algs/ags/solver.cc#L214

This line calls `UpdateAllH`, passing a `set<Interval*, ags::CompareIntervals>::iterator` as the argument. However,

https://github.com/stevengj/nlopt/blob/b22a9f7f65e589022bc9a775ad9977c792fab123/src/algs/ags/solver.hpp#L82

`UpdateAllH` is defined to take a `set<Interval*>::iterator`, i.e. a `set<Interval*, std::less<Interval*>>::iterator`. These are technically different and incompatible iterators, although without debug iterators they do collapse to an underlying `__rb_tree_iterator<Interval*>`.

I don't 100% understand the code, but if I had to guess, the proper fix is to change the signature of `UpdateAllH` to take a `std::set<Interval*, ags::CompareIntervals>::iterator` (it may make sense to define a typedef/using for that set type). If it needs to support iterators to differently compared sets, presumably some kind of lightweight templating is needed.

### How to reproduce the issue?

```shell
Try to build nlopt with `-D_GLIBCXX_DEBUG`. Observe this error:


/home/sliedes/foo/source/src/algs/ags/solver.cc:214:16: error: no viable conversion from '_Safe_iterator<[...], set<[...], ags::CompareIntervals, [...]>, [...]>' to '_Safe_iterator<[...], set<[...], std::less<ags::Interval *>, [...]>, [...]>'
  214 |     UpdateAllH(insRes.first);
      |                ^~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:627:7: note: candidate constructor not viable: no known conversion from '__gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, ags::CompareIntervals>>' to 'const _Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &' for 1st argument
  627 |       _Safe_iterator(const _Safe_iterator& __x) _GLIBCXX_NOEXCEPT
      |       ^              ~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:634:7: note: candidate constructor not viable: no known conversion from '__gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, ags::CompareIntervals>>' to '_Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &&' for 1st argument
  634 |       _Safe_iterator(_Safe_iterator&&) = default;
      |       ^              ~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:643:2: note: candidate template ignored: could not match 'set<[...], std::less<ags::Interval *>, [...]>' against 'set<[...], ags::CompareIntervals, [...]>'
  643 |         _Safe_iterator(
      |         ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:444:7: note: candidate function
  444 |       operator _Iterator() const _GLIBCXX_NOEXCEPT { return *this; }
      |       ^
/home/sliedes/foo/source/src/algs/ags/solver.hpp:82:48: note: passing argument to parameter here
   82 |   void UpdateAllH(std::set<Interval*>::iterator);
      |                                                ^
/home/sliedes/foo/source/src/algs/ags/solver.cc:279:16: error: no viable conversion from '_Safe_iterator<[...], set<[...], ags::CompareIntervals, [...]>, [...]>' to '_Safe_iterator<[...], set<[...], std::less<ags::Interval *>, [...]>, [...]>'
  279 |     UpdateAllH(insResult.first);
      |                ^~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:627:7: note: candidate constructor not viable: no known conversion from '__gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, ags::CompareIntervals>>' to 'const _Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &' for 1st argument
  627 |       _Safe_iterator(const _Safe_iterator& __x) _GLIBCXX_NOEXCEPT
      |       ^              ~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:634:7: note: candidate constructor not viable: no known conversion from '__gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, ags::CompareIntervals>>' to '_Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &&' for 1st argument
  634 |       _Safe_iterator(_Safe_iterator&&) = default;
      |       ^              ~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:643:2: note: candidate template ignored: could not match 'set<[...], std::less<ags::Interval *>, [...]>' against 'set<[...], ags::CompareIntervals, [...]>'
  643 |         _Safe_iterator(
      |         ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:444:7: note: candidate function
  444 |       operator _Iterator() const _GLIBCXX_NOEXCEPT { return *this; }
      |       ^
/home/sliedes/foo/source/src/algs/ags/solver.hpp:82:48: note: passing argument to parameter here
   82 |   void UpdateAllH(std::set<Interval*>::iterator);
      |                                                ^
/home/sliedes/foo/source/src/algs/ags/solver.cc:280:16: error: no viable conversion from '_Safe_iterator<[...], set<[...], ags::CompareIntervals, [...]>, [...]>' to '_Safe_iterator<[...], set<[...], std::less<ags::Interval *>, [...]>, [...]>'
  280 |     UpdateAllH(--insResult.first);
      |                ^~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:627:7: note: candidate constructor not viable: no known conversion from '_Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *, CompareIntervals>>' to 'const _Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &' for 1st argument
  627 |       _Safe_iterator(const _Safe_iterator& __x) _GLIBCXX_NOEXCEPT
      |       ^              ~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:634:7: note: candidate constructor not viable: no known conversion from '_Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *, CompareIntervals>>' to '_Safe_iterator<_Rb_tree_const_iterator<Interval *>, set<Interval *>> &&' for 1st argument
  634 |       _Safe_iterator(_Safe_iterator&&) = default;
      |       ^              ~~~~~~~~~~~~~~~~
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:643:2: note: candidate template ignored: could not match 'set<[...], std::less<ags::Interval *>, [...]>' against 'set<[...], ags::CompareIntervals, [...]>'
  643 |         _Safe_iterator(
      |         ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:444:7: note: candidate function
  444 |       operator _Iterator() const _GLIBCXX_NOEXCEPT { return *this; }
      |       ^
/home/sliedes/foo/source/src/algs/ags/solver.hpp:82:48: note: passing argument to parameter here
   82 |   void UpdateAllH(std::set<Interval*>::iterator);
      |                                                ^
/home/sliedes/foo/source/src/algs/ags/solver.cc:359:11: error: cannot cast 'const std::set<Interval *>::iterator' (aka 'const _Safe_iterator<_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, std::less<ags::Interval *>, std::allocator<ags::Interval *>>>') to its private base class 'const _Self' (aka 'const _Rb_tree_const_iterator<ags::Interval *>')
  359 |     while(rightIterator != mSearchInformation.end() && (*rightIterator)->pl.idx < pInterval->pl.idx)
      |           ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:131:7: note: declared private here
  131 |     : private _Iterator,
      |       ^~~~~~~~~~~~~~~~~
/home/sliedes/foo/source/src/algs/ags/solver.cc:361:9: error: cannot cast 'const std::set<Interval *>::iterator' (aka 'const _Safe_iterator<_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, std::less<ags::Interval *>, std::allocator<ags::Interval *>>>') to its private base class 'const _Self' (aka 'const _Rb_tree_const_iterator<ags::Interval *>')
  361 |     if (rightIterator != mSearchInformation.end() && (*rightIterator)->pl.idx >= pInterval->pl.idx)
      |         ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:131:7: note: declared private here
  131 |     : private _Iterator,
      |       ^~~~~~~~~~~~~~~~~
/home/sliedes/foo/source/src/algs/ags/solver.cc:370:11: error: cannot cast 'const std::set<Interval *>::iterator' (aka 'const _Safe_iterator<_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, std::less<ags::Interval *>, std::allocator<ags::Interval *>>>') to its private base class 'const _Self' (aka 'const _Rb_tree_const_iterator<ags::Interval *>')
  370 |     while(leftIterator != mSearchInformation.begin() && (*leftIterator)->pl.idx < pInterval->pl.idx)
      |           ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:131:7: note: declared private here
  131 |     : private _Iterator,
      |       ^~~~~~~~~~~~~~~~~
/home/sliedes/foo/source/src/algs/ags/solver.cc:372:9: error: cannot cast 'const std::set<Interval *>::iterator' (aka 'const _Safe_iterator<_Rb_tree_const_iterator<ags::Interval *>, std::set<ags::Interval *, std::less<ags::Interval *>, std::allocator<ags::Interval *>>>') to its private base class 'const _Self' (aka 'const _Rb_tree_const_iterator<ags::Interval *>')
  372 |     if (leftIterator != mSearchInformation.begin() && (*leftIterator)->pl.idx >= pInterval->pl.idx)
      |         ^
/nix/store/jyphzc03k61s72s290l97npk0p9daspz-gcc-14.3.0/include/c++/14.3.0/debug/safe_iterator.h:131:7: note: declared private here
  131 |     : private _Iterator,
      |       ^~~~~~~~~~~~~~~~~
7 errors generated.
make[2]: *** [CMakeFiles/nlopt.dir/build.make:685: CMakeFiles/nlopt.dir/src/algs/ags/solver.cc.o] Error 1
make[2]: Leaving directory '/home/sliedes/foo/source/build'
make[1]: *** [CMakeFiles/Makefile2:160: CMakeFiles/nlopt.dir/all] Error 2
make[1]: Leaving directory '/home/sliedes/foo/source/build'
make: *** [Makefile:166: all] Error 2
```

### Version

nlopt 2.10.0

### Operating System

Linux

### Installation media

from source

### Additional Context

_No response_

--- a/src/algs/ags/solver.cc
+++ b/src/algs/ags/solver.cc
@@ -341,7 +341,7 @@ void NLPSolver::UpdateH(double newValue, int index)
   }
 }
 
-void NLPSolver::UpdateAllH(std::set<Interval*>::iterator iterator)
+void NLPSolver::UpdateAllH(std::set<Interval*, ags::CompareIntervals>::iterator iterator)
 {
   Interval* pInterval = *iterator;
   if (pInterval->pl.idx < 0)

--- a/src/algs/ags/solver.hpp
+++ b/src/algs/ags/solver.hpp
@@ -79,7 +79,7 @@ protected:
   void InitDataStructures();
   void ClearDataStructures();
 
-  void UpdateAllH(std::set<Interval*>::iterator);
+  void UpdateAllH(std::set<Interval*, ags::CompareIntervals>::iterator);
   void UpdateH(double newValue, int index);
   double CalculateR(Interval*) const;
   double GetNextPointCoordinate(Interval*) const;

